File: __iOS_BACKUP__/ShapeByte/shared/ui/dimensions/DimensionProvider.swift
//
//  DimensionProvider.swift
//  ShapeByte
//
//  Created by Lang, Stefan [RTL Tech] on 14.08.24.
//

import Foundation

// TODO: test
class DimensionProvider {
    let defaultSize: CGSize
    lazy var deviceSizeCategory: DeviceSizeCategory = deviceSizeCategoryProvider.sizeCategory

    var screenSize: CGSize {
        return deviceSizeCategory.size
    }

    private let deviceSizeCategoryProvider: DeviceSizeCategoryProvider

    init(
        defaultSize: CGSize = DeviceSizeCategoryProvider.iPhoneSizeBounds.medium.size,
        deviceSizeCategoryProvider: DeviceSizeCategoryProvider
    ) {
        self.defaultSize = defaultSize
        self.deviceSizeCategoryProvider = deviceSizeCategoryProvider
    }

    func withDimensionalAspect(height: CGFloat, max: CGFloat? = nil) -> CGFloat {
        let heightFactor = screenSize.height / defaultSize.height
        let result = height * heightFactor

        guard let maxNotNil = max else {
            return result
        }

        return min(maxNotNil, result)
    }
}

File: foundation/presentation/contract/src/commonMain/kotlin/de/stefan/lang/foundation/presentation/contract/dimension/DimensionProvider.kt
package de.stefan.lang.foundation.presentation.contract.dimension

import de.stefan.lang.foundation.core.contract.devicesize.DeviceSizeCategory
import de.stefan.lang.foundation.core.contract.devicesize.DeviceSizeCategoryProvider
import de.stefan.lang.foundation.core.contract.devicesize.Size
import kotlin.math.min

public class DimensionProvider(
    private val deviceSizeCategoryProvider: DeviceSizeCategoryProvider,
) {
    public val screenSize: Size
        get() = deviceSizeCategory.size

    public val deviceSizeCategory: DeviceSizeCategory
        get() = deviceSizeCategoryProvider.sizeCategory

    private val defaultSize: Size by lazy {
        deviceSizeCategoryProvider
            .sizeCategories()
            .filterIsInstance<DeviceSizeCategory.Medium>().first().size
    }

    /*
     * For Obj-C interop
     * */
    public fun withDimensionalAspect(height: Float): Float {
        return withDimensionalAspect(height, Float.MAX_VALUE)
    }

    public fun withDimensionalAspect(height: Float, max: Float): Float {
        val heightFactor = screenSize.height / defaultSize.height
        val result = height * heightFactor

        val retVal = min(max, result)
        return retVal
    }
}
